docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,210 @@
+
+  Claim: Maintaining and improving a custom RPC protocol is a substantial
+  project and not directly relevant to the requirements of consensus. We would
+  be better served by depending on a well-maintained open-source library like
+  gRPC.
+
+- Question 2b: Improve the socket protocol and deprecate/remove gRPC?
+
+  Claim: If we find meaningful advantages to maintaining our own custom RPC
+  protocol in Tendermint, we should treat it as a first-class project within
+  the core and invest in making it good enough that we do not require other
+  options.
+
+**One important consideration** when discussing these questions is that _any
+outcome which includes keeping the socket protocol will have eventual migration
+impacts for clients_ regardless. To fix the limitations of the socket protocol
+as it is currently designed will require making _breaking changes_ to the
+protocol.  So, while we may put off a migration cost for clients by retaining
+the socket protocol in the short term, we will eventually have to pay those
+costs to fix the problems in its current design.
+
+## Detailed Design
+
+1. If we choose to standardize on gRPC, the main work in in Tendermint core
+   will be removing and cleaning up the code for the socket client and server.
+
+   Besides the code cleanup, we will also need to clearly document a
+   deprecation schedule, and invest time in making the migration easier for
+   applications currently using the socket protocol.
+
+   > **Point for discussion:** Migrating from the socket protocol to gRPC
+   > should mostly be a plumbing change, as long as we do it during a release
+   > in which we are not making other breaking changes to ABCI. However, the
+   > effort may be more or less depending on how gRPC integration works in the
+   > application's implementation language, and would have to be sure networks
+   > have plenty of time not only to make the change but to verify that it
+   > preserves the function of the network.
+   >
+   > What questions should we be asking node operators and application
+   > developers to understand the migration costs better?
+
+2. If we choose to keep only the socket protocol, we will need to follow up
+   with a more detailed design for extending and upgrading the protocol to fix
+   the existing performance and operational issues with the protocol.
+
+3. If we choose to keep both options, we will still need to do all the work of
+   (2), but the gRPC implementation should not require any immediate changes.
+
+
+## Consequences
+
+- **Standardize on gRPC**
+
+   - ✅ Addresses existing performance and operational issues.
+   - ✅ Replaces custom code with a well-maintained widely-used library.
+   - ✅ Aligns well with Cosmos SDK, which already uses gRPC extensively.
+   - ✅ Aligns well priv validator interface, for which the socket protocol is already deprecated for gRPC.
+   - ❓ Applications will be hard to implement in a language without gRPC support.
```

Comment:
Is there an actual example of such language?

Reply:
> Is there an actual example of such language?

The current list of officially supported languages includes C#, C++, Dart, Go, Java, Kotlin, Node, Objective-C, PHP, Python, and Ruby. There are multiple community-contributed implementations for Rust, and other languages (e.g., OCaml, Scala, etc.) seem to have reasonable support.

In practical terms, I think that covers most of the languages anyone is likely to use for production applications, and it looks like community implementations of gRPC are probably available for many others. That's why I marked this a ❓ rather than a ⛔, since I think any language not covered is unlikely to have much impact in the space.

Reply:
How certain are we on the quality and maintenance guarantees on the best implementations for Rust? (which is currently the obvious language of choice besides Go)

Reply:
> How certain are we on the quality and maintenance guarantees on the best implementations for Rust? (which is currently the obvious language of choice besides Go)

I don't know enough to answer. 

My intuition is that if someone is going to have to maintain a wire protocol implementation either way, it would probably still be better for that to be gRPC than the custom socket protocol.

Reply:
> How certain are we on the quality and maintenance guarantees on the best implementations for Rust? (which is currently the obvious language of choice besides Go)

It might be reasonable to do a bit more diligence, but the standard validator stack includes a signing service (tmkms) that is moving to use gRPC (for our privval interface) so I think practically speaking it's _fine_.

Reply:
> My intuition is that if someone is going to have to maintain a wire protocol implementation either way, it would probably still be better for that to be gRPC than the custom socket protocol.

Is this really true given the custom socket protocol should be *much* simpler than an implementation of grpc?

It might be valuable to try and collect some feedback on grpc implementation quality in non official languages. Hopefully it will be the case that by the time anyone wants to develop an ABCI app in another language there will be a way to use grpc from that language. I would hate to see someone want to experiment with ABCI apps in a new lang and not be able to because of the grpc blocker, but that will probably be pretty rare (I guess worse case they can prob always use FFI to a lang with a grpc impl on the server side ? )

Reply:
Agreed that gRPC is a more complex protocol than the socket protocol as it is currently designed. On the other hand, most of the likely languages for app development already have good (either official or unofficial) gRPC implementations. 

The important question here, I think, is what to do about the long tail of other options. Broadly I claim we should not worry too much about that long tail, on the basis that it's unlikely to generate any meaningful impact for the ecosystem.

Under the covers, gRPC is around the same level of complexity for an implementor as websockets. The substrate is stock HTTP/2, and the message formats are protocol buffers. We already require anyone building against ABCI (of any version) to have protobuf support, and HTTP is usually an out-of-box library feature.

Despite being more work, gRPC is also widely used for other purposes (beyond ABCI) throughout the Tendermint ecosystem, so an investment in it would have broader potential returns for the application. It _does_ mean that a developer using some unusual language would have to do more short-term work to get something running, but I'm not sure we should invest a lot in the corner case of "weird languages, toy problems".

Perhaps more importantly, if the goal is to "pick one (and only one) mechanism for out-of-process apps" (which is the decision this ADR is proposing), I think developers who _are_ using a well-supported gRPC language will have a better experience (and matter more to our impact).